### PR TITLE
Initial terraform: Send S3 events on a SQS queue

### PR DESCRIPTION
**Pull request by arielshaqed**

### CODE REVIEW

File: tf/s3_events.tf [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  bucket = var.bucket
+}
+
+# This should properly be a FIFO queue, to avoid mixing up Create and Delete
+# operations.  However AWS says:
+#
+# error putting S3 Bucket Notification Configuration: InvalidArgument: FIFO
+#     SQS queues are not supported.
+resource "aws_sqs_queue" "s3_events_queue" {
+  name = "terminus-s3-events-queue"
```
Comment (johnnyaug) [inline tf/s3_events.tf:13]: Can terminus be used for more than one bucket? If so, the queue should have a unique name for every bucket

File: tf/s3_events.tf [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# AWS_DEFAULT_REGION=eu-central-1 terraform plan -var bucket=treeverse-ariels-test
```
Comment (johnnyaug) [inline tf/s3_events.tf:1]: /dd

File: tf/s3_events.tf [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  # ReceiveMessage is a long poll.
+  receive_wait_time_seconds = 20
+  
+  # Could use a dead-letter queue: no reordering issues are expected for S3
+  # events, as all objects must have distinct names to be counted correctly.
+  # This may be an issue as a poisoned message blocks processing.  Will need
+  # to monitor the queue.
+  #
+  # If we *do* want to work for objects with distinct names, the ordering of
+  # messages becomes critical, and dead-letter queues cannot be used at all.
```
Comment (johnnyaug) [inline tf/s3_events.tf:27]: Very technical and not clear to me, not sure this is the place for this

**Review by johnnyaug — CHANGES_REQUESTED** [review]
Nice and clean, thanks!

**Review by arielshaqed — COMMENTED** [review]
Thanks!  All good comments.

**Comment by arielshaqed** [comment]
SORRY!! Changed wrong branch; please hold on reviewing for a while.

**Comment by arielshaqed** [comment]
Thanks -- better now I hope.

File: tf/s3_events.tf (hunk 1 above)
Comment (arielshaqed) [reply tf/s3_events.tf:30]: Yes it can!  Modified to support multiple buckets.

**Review by arielshaqed — COMMENTED** [review]
Thanks!

I learned a lot about HCL and about Terraform when adding multiple bucket support, so please do review changes with some care.

## The good

* The policy is now structured in HCL rather than plain JSON.  So bucket ARN interpolation actually works out nicely.

## The bad

I have to use two distinct iteration mechanisms:

* `for_each` meta-arguments to create multiple resource or data blocks, one for each bucket.
* `for` expressions (like functional "map") to create the list of allowed Source ARNs inside the Condition on the IAM policy.

Is there a better way?

**Review by johnnyaug — APPROVED** [review]

**Comment by johnnyaug** [comment]
Looks really good!

**Comment by arielshaqed** [comment]
Thanks!

Pulling...

**Pull request merged**